### PR TITLE
Implement list archives

### LINKED ISSUE #489: [API] get the list of archives
## Feature Request

/archives/all: Get the list of archives

See [here](https://app.swaggerhub.com/apis/ChaosMesh/chaos-server/1.0.0#trial) for details

### PR DESCRIPTION
**Pull request by yeya24**
Signed-off-by: yeya24 <yb532204897@gmail.com>

### What problem does this PR solve? 

Fixes #489

### CODE REVIEW

File: pkg/store/experiment/experiment.go [hunk 1]
```diff
@@ -32,18 +33,77 @@ type experimentStore struct {
 }
 
 // TODO: implement core.EventStore interface
```
Comment (cwen0) [inline pkg/store/experiment/experiment.go:35]: We can remove this TODO comment

File: pkg/store/experiment/experiment_test.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
+package experiment
```
[the remark this answers is not in the document]
Comment (yeya24) [reply pkg/store/experiment/experiment_test.go:1]: DONE

File: pkg/store/experiment/experiment.go [hunk 3]
```diff
@@ -32,18 +33,77 @@ type experimentStore struct {
...
-func (e *experimentStore) Find(context.Context, int64) (*core.ArchiveExperiment, error) {
-	return nil, nil
+
+func (e *experimentStore) ListMeta(_ context.Context, kind, ns, name string) ([]*core.ArchiveExperimentMeta, error) {
+	archives := make([]*core.ArchiveExperimentMeta, 0)
+	query, args := constructQueryArgs(kind, ns, name)
+
+	// List all experiment metas
+	if len(args) == 0 {
+		if err := e.db.Table("archive_experiments").Find(&archives).Error; err != nil && !gorm.IsRecordNotFoundError(err) {
```
[the remark this answers is not in the document]
Comment (yeya24) [reply pkg/store/experiment/experiment.go:61]: If the table is not set, gorm will try to query from table `archive_experiment_meta`. So I have to set this manually.

File: pkg/apiserver/archive/experiment.go [hunk 4]
```diff
@@ -48,11 +74,48 @@ func NewService(
...
+// @Failure 500 {object} utils.APIError
+func (s *Service) listExperiments(c *gin.Context) {
+	kind := c.Query("kind")
+	name := c.Query("name")
+	ns := c.Query("namespace")
+
+	metas, err := s.archive.ListMeta(context.TODO(), kind, ns, name)
+	if err != nil {
+		c.Status(http.StatusInternalServerError)
+		_ = c.Error(utils.ErrInternalServer.NewWithNoMessage())
```
[the remark this answers is not in the document]
Comment (yeya24) [reply pkg/apiserver/archive/experiment.go:101]: DONE

File: pkg/apiserver/archive/experiment.go [hunk 5]
```diff
@@ -14,15 +14,41 @@
...
 	"github.com/pingcap/chaos-mesh/pkg/core"
 
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	"github.com/pingcap/chaos-mesh/pkg/config"
 )
 
+var (
+	// TODO(yeya24): remove this after moving this to one place
+	kinds = map[string]v1alpha1.ChaosList{
```
[the remark this answers is not in the document]
Comment (yeya24) [reply pkg/apiserver/archive/experiment.go:33]: DONE

File: pkg/apiserver/archive/experiment.go [hunk 6]
```diff
@@ -14,15 +14,41 @@
...
+		v1alpha1.KindPodChaos:     &v1alpha1.PodChaosList{},
+		v1alpha1.KindNetworkChaos: &v1alpha1.NetworkChaosList{},
+		v1alpha1.KindStressChaos:  &v1alpha1.StressChaosList{},
+		v1alpha1.KindIOChaos:      &v1alpha1.IoChaosList{},
+		v1alpha1.KindKernelChaos:  &v1alpha1.KernelChaosList{},
+		v1alpha1.KindTimeChaos:    &v1alpha1.TimeChaosList{},
+	}
+)
+
+type ExperimentMeta struct {
```
[the remark this answers is not in the document]
Comment (yeya24) [reply pkg/apiserver/archive/experiment.go:43]: DONE

File: pkg/apiserver/archive/experiment.go [hunk 7]
```diff
@@ -14,15 +14,29 @@
 package archive
 
 import (
-	"github.com/gin-gonic/gin"
+	"context"
+	"net/http"
+	"time"
 
+	"github.com/gin-gonic/gin"
```
Comment (cwen0) [inline pkg/apiserver/archive/experiment.go:21]: group these

File: pkg/apiserver/archive/experiment.go [hunk 8]
```diff
@@ -14,15 +14,29 @@
...
+	"github.com/pingcap/chaos-mesh/pkg/apiserver/utils"
 	"github.com/pingcap/chaos-mesh/pkg/core"
 
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	"github.com/pingcap/chaos-mesh/pkg/config"
 )
 
+// ArchiveExperimentMeta defines the meta data of archived experiments
+type ArchiveExperimentMeta struct {
```
[the remark this answers is not in the document]
Comment (cwen0) [reply pkg/apiserver/archive/experiment.go:31]: How about use `archive.ArchiveExperimentMeta` directly?

**Comment by yeya24** [comment]
> How about use archive.ArchiveExperimentMeta directly?

I tried, but this is not possible because `archive.ArchiveExperimentMeta` includes `gorm.Model` as well so it will return this struct 

```
type Model struct {
	ID        uint `gorm:"primary_key"`
	CreatedAt time.Time
	UpdatedAt time.Time
	DeletedAt *time.Time `sql:"index"`
}
```

File: pkg/apiserver/archive/experiment.go (hunk 7 above)
Comment (yeya24) [reply pkg/apiserver/archive/experiment.go:21]: Done

**Comment by cwen0** [comment]
> > How about use archive.ArchiveExperimentMeta directly?
> 
> I tried, but this is not possible because `archive.ArchiveExperimentMeta` includes `gorm.Model` as well so it will return this struct
> 
> ```
> type Model struct {
> 	ID        uint `gorm:"primary_key"`
> 	CreatedAt time.Time
> 	UpdatedAt time.Time
> 	DeletedAt *time.Time `sql:"index"`
> }
> ```

Why we can't return `gorm.Model`? If we return it, we can use the ID to get details of the archive experiment.

**Comment by yeya24** [comment]
Ok. But that’s not part of the API designed before.   Will update that tomorrow if that ok

**Comment by yeya24** [comment]
> > > How about use archive.ArchiveExperimentMeta directly?
> > 
> > 
> > I tried, but this is not possible because `archive.ArchiveExperimentMeta` includes `gorm.Model` as well so it will return this struct
> > ```
> > type Model struct {
> > 	ID        uint `gorm:"primary_key"`
> > 	CreatedAt time.Time
> > 	UpdatedAt time.Time
> > 	DeletedAt *time.Time `sql:"index"`
> > }
> > ```
> 
> Why we can't return `gorm.Model`? If we return it, we can use the ID to get details of the archive experiment.

Updated. I include the id field in ArchiveExperimentMeta. I still don't want to use `archive.ArchiveExperimentMeta` because it will include other times.

```
  {
    "ID": 2406,
    "CreatedAt": "2020-05-26T10:15:20.057569-04:00",
    "UpdatedAt": "2020-05-26T10:15:20.057569-04:00",
    "DeletedAt": null,
    "Name": "tipocket-cdc-kafka-tipocket-cdc-kafka-downstream-tikv-2-time-chaos",
    "Namespace": "tipocket-cdc-kafka",
    "Kind": "TimeChaos",
    "Action": "",
    "StartTime": "2020-05-14T18:56:42-04:00",
    "FinishTime": "2020-05-14T18:58:01-04:00"
  },
  {
    "ID": 2407,
    "CreatedAt": "2020-05-26T10:15:20.058333-04:00",
    "UpdatedAt": "2020-05-26T10:15:20.058333-04:00",
    "DeletedAt": null,
    "Name": "tipocket-cdc-kafka-tipocket-cdc-kafka-downstream-tidb-0-time-chaos",
    "Namespace": "tipocket-cdc-kafka",
    "Kind": "TimeChaos",
    "Action": "",
    "StartTime": "2020-05-14T18:56:41-04:00",
    "FinishTime": "2020-05-14T18:59:09-04:00"
  },
```

**Comment by yeya24** [comment]
> > > How about use archive.ArchiveExperimentMeta directly?
> > 
> > 
> > I tried, but this is not possible because `archive.ArchiveExperimentMeta` includes `gorm.Model` as well so it will return this struct
> > ```
> > type Model struct {
> > 	ID        uint `gorm:"primary_key"`
> > 	CreatedAt time.Time
> > 	UpdatedAt time.Time
> > 	DeletedAt *time.Time `sql:"index"`
> > }
> > ```
> 
> Why we can't return `gorm.Model`? If we return it, we can use the ID to get details of the archive experiment.

Updated, please take a look @cwen0

**Pull request merged**